sql/hints: add per-node cache for system.statement_hints

DrewKimball opened the pull request:
#### sql/hints: change StatementHint protobuf struct to union

This commit makes a few changes to the protobuf message used for
serializing statement hints in the `system.statement_hints` table to
prepare for future commits/PRs. No release note or migrations are
necessary, since nothing is currently store nor read from the table.

Epic: None

Release note: None

#### sql/hints: add per-node cache for system.statement_hints

This commit adds a per-node cache for the `system.statement_hints` table,
similar to the table stats cache. The cache tracks the set of all statement
fingerprint hashes that have hints in a map kept up-to-date via rangefeed.
This allows queries to quickly determine whether they have hints, minimizing
the overhead for unhinted queries. Hints are kept in a separate
limited-capacity LRU cache. The LRU cache size can be configured by setting
the `sql.hints.statement_hints_cache_size` cluster setting.

Fixes #148160

Release note: None

#### sql/hints: wait for initial scan during server startup

Server startup will now block until the initial scan of the
`system.statement_hints` table for the hints cache has finished.
This ensures that the node will not accept queries until `hintedHashes`
is initialized, which means that queries without hints will not have
to perform system table reads.

It is possible for the initial scan to fail, in which case server startup
will unblock while the scan restarts. In this rare error scenario,
queries will have to read from the system table on the first execution
even if they don't have hints until the scan finishes. This is
considered preferable to blocking startup across retries, since the
`hintedHashes` map is only an optimization.

Epic: None

Release note: None

#### sql/hints: add tests for statement hints cache

This commit adds a set of tests for the behavior of the hint cache.
The tests verify that the cache is correctly populated in response
to existing hints as well as new insertions and deletions.

Epic: None

Release note: None

DrewKimball committed:
sql/hints: change StatementHint protobuf struct to union

This commit makes a few changes to the protobuf message used for
serializing statement hints in the `system.statement_hints` table to
prepare for future commits/PRs. No release note or migrations are
necessary, since nothing is currently store nor read from the table.

Epic: None

Release note: None

DrewKimball committed:
sql/hints: add per-node cache for system.statement_hints

This commit adds a per-node cache for the `system.statement_hints` table,
similar to the table stats cache. The cache tracks the set of all statement
fingerprint hashes that have hints in a map kept up-to-date via rangefeed.
This allows queries to quickly determine whether they have hints, minimizing
the overhead for unhinted queries. Hints are kept in a separate
limited-capacity LRU cache. The LRU cache size can be configured by setting
the `sql.hints.statement_hints_cache_size` cluster setting.

Fixes #148160

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153995)
<!-- Reviewable:end -->

DrewKimball commented:
@michae2 I plan on adding tests + possibly make server startup wait on the initial scan before marking this PR as ready, but thought it would be a good idea to let you see the implementation earlier. I decided to go back to an earlier idea, to keep track of the number of hints associated with each hash. This is nice because it avoids the need to directly read from the DB at all, except to actually fetch the hints themselves.

It turns out that `rangefeedcache.Watcher` is designed for this type of use-case: https://github.com/cockroachdb/cockroach/blob/1a26fe72ade9ecc03db10fff21cc531f0b8087cf/pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go#L31-L38
Where an in-memory data structure is updated by rangefeed events. There are some existing places where we maintain a cache using only the rangefeed (rather than having the rangefeed events trigger a DB read), for example, [here](https://github.com/cockroachdb/cockroach/blob/1a26fe72ade9ecc03db10fff21cc531f0b8087cf/pkg/server/tenantsettingswatcher/watcher.go#L150).

I did also play around with having rangefeed events trigger async DB reads that populate the map, but there is some complexity around handling failures and making sure race conditions don't clobber newer values that convinced me to go with the other approach. Happy to explore this more / discuss further depending on what you think.

DrewKimball committed:
sql/hints: add tests for statement hints cache

This commit adds a set of tests for the behavior of the hint cache.
The tests verify that the cache is correctly populated in response
to existing hints as well as new insertions and deletions.

Epic: None

Release note: None

michae2 reviewed (commented):
Looking quite good so far. I'm still reading...

> @michae2 reviewed 2 of 2 files at r1, 9 of 9 files at r3, 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-Oa2JG4H5EIpzydIRa7v:b9k0fnr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/hints/hint_cache.go` line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa25y4EA-3fwRuNakTj:-Oa25y4F6pCtvuZPLnX9:b2nxeti) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L131)):*
> ```Go
> 		return err
> 	}
> 	statementHintsHashIndexPrefix := keys.SystemSQLCodec.IndexPrefix(
> ```

I think codec needs to be a parameter of start, because we probably want this hints table to be tenant-specific in the same way the stats cache is.

Also, did we decide whether statement_hints will be replicated and/or available to read-only tenants? If statement_hints will be available to read-only tenants via external row data, then these spans need to use the external tenant ID and table ID.

(It's possible the stats cache is doing this wrong... I believe table_statistics is set up to use external row data in pkg/sql/catalog/replication/reader_catalog.go but I'm not sure we adjust the rangefeed in stats_cache for external row data...)
___
*[`pkg/sql/hints/hint_cache.go` line 159 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa2DFgl41I2Q1ZgT_jm:-Oa2DFgmDRNi_s-Nzoka:bb9k0j5) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L159)):*
> ```Go
> 	ctx context.Context, kv *kvpb.RangeFeedValue,
> ) (*bufferEvent, bool) {
> 	hash, err := decodeHashFromStatementHintsKey(keys.SystemSQLCodec, kv.Key, &c.datumAlloc)
> ```

Same comment: probably shouldn't be the system codec if statement_hints is per-tenant.
___
*[`pkg/sql/hints/hint_cache.go` line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa2HghRBeQJebj3Zs_f:-Oa2HghRBeQJebj3Zs_g:bssg135) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L265)):*
> ```Go
> 		// If the hintedHashes map has not yet been initialized, statements must
> 		// unconditionally check hintCache.
> 		if _, ok := c.mu.hintedHashes[statementHash]; !ok {
> ```

Maybe defensively we should also `return nil` if the number of hintedHashes is <= 0.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
Thanks for taking a look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-Oa75k3k7sMkemH7xsl9:bubxju8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/hints/hint_cache.go` line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa25y4EA-3fwRuNakTj:-Oa6wKGSAP-IE5YHwKp2:b7vpvdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L131)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think codec needs to be a parameter of start, because we probably want this hints table to be tenant-specific in the same way the stats cache is.

Also, did we decide whether statement_hints will be replicated and/or available to read-only tenants? If statement_hints will be available to read-only tenants via external row data, then these spans need to use the external tenant ID and table ID.

(It's possible the stats cache is doing this wrong... I believe table_statistics is set up to use external row data in pkg/sql/catalog/replication/reader_catalog.go but I'm not sure we adjust the rangefeed in stats_cache for external row data...)
</blockquote></details>

That's a great catch. Fixed it, and added a test.
___
*[`pkg/sql/hints/hint_cache.go` line 159 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa2DFgl41I2Q1ZgT_jm:-Oa6y_qTAy07TdgPCtKy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L159)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Same comment: probably shouldn't be the system codec if statement_hints is per-tenant.
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa2HghRBeQJebj3Zs_f:-Oa75VXz96L99nFmLtrU:bxeux1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L265)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Maybe defensively we should also `return nil` if the number of hintedHashes is <= 0.
</blockquote></details>

I'd argue the current state is more defensive, since it means we'll fall back to the LRU cache and read any hints from the table. What do you think?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Great work! I only some minor comments and some questions.

> @yuzefovich reviewed 1 of 2 files at r1, 1 of 9 files at r4, 12 of 12 files at r7, 8 of 8 files at r8, 1 of 1 files at r9, 5 of 5 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-OaRn3A76hK91i-o-3DD:by8cebe)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/hints/hint_cache_test.go` line 36 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRkM7zBH7IauZTyBS0:-OaRkM7zBH7IauZTyBS1:b-bj34qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache_test.go#L36)):*
> ```Go
> 	ctx := context.Background()
> 	srv, db, _ := serverutils.StartServer(t, base.TestServerArgs{
> 		DefaultTestTenant: base.TestIsSpecificToStorageLayerAndNeedsASystemTenant,
> ```

nit: why are some of these tests explicitly choose to not use randomized test tenant? I'd expect that things should just work, modulo perhaps a more complicated setup where we need to do something (like updating SystemOnly cluster settings) against the system tenant explicitly.
___
*[`pkg/sql/hints/hint_cache_test.go` line 257 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRmaZVCshVJ6V665Hr:-OaRmaZVCshVJ6V665Hs:bgxdl6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache_test.go#L257)):*
> ```Go
> 	defer tc.Stopper().Stop(ctx)
> 
> 	// Use node 0 as the primary node for the cache
> ```

nit: missing period.
___
*[`pkg/sql/hints/hint_cache.go` line 237 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRisPPDv8GWU7PX5VU:-OaRisPPDv8GWU7PX5VV:b-xx0jnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dd47b4ac71506d17e1b811332d5a3009854c477/pkg/sql/hints/hint_cache.go#L237)):*
> ```Go
> 		c.initialScanChan <- err
> 		close(c.initialScanChan)
> 		c.initialScanChan = nil
> ```

Just double-checking that we expect that `onError` and `onUpdate` handlers are called from the same goroutine, right? (I think so based on a quick look at the code.)
___
*[`pkg/sql/hints/hint_cache.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlqS1-MLC9QbnZErZ:-OaRlqS2A4HOqGS_2lzf:b-uha57q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L40)):*
> ```Go
> 
> // StatementHintsCache allows statements to access their external hints stored
> // in the system.statement_hints table (and determine whether they have hints)
> ```

Should we somehow handle the internal queries in a special way? Do we foresee ever adding hints for them? It sounds like with the current approach we expect them to not be found in `hintedHashes` (right?), so maybe the overhead of that check is negligible.
___
*[`pkg/sql/hints/hint_cache.go` line 444 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlSj8AgE0ZKje3TK5:-OaRlSj8AgE0ZKje3TK6:bcghhvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L444)):*
> ```Go
> 	var res []StatementHint
> 	for i := range entry.hints {
> 		if entry.fingerprints[i] == statementFingerprint {
> ```

I have a somewhat general question. Do we want for hints to apply across databases? It seems like right now if we happen to have a hash match and then the stmt fingerprint matches, we'll apply all hints, regardless of the DB, and perhaps that's intended.
___
*[`pkg/sql/hints/hint_cache.go` line 55 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRbNdoATqph43n836D:-OaRbNdp0wjMBrcrVZUN:b-4fcobr) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L55)):*
> ```Go
> 		// to quickly determine whether there are any hints for a given statement,
> 		// ensuring that overhead is minimized for non-hinted statements.
> 		hintedHashes map[int64]int
> ```

nit: it'd be good to mention what are the "key" and the "value" in this map.
___
*[`pkg/sql/hints/hint_cache.go` line 83 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaR_qFc5UdhZ3MSYm3G:-OaR_qFc5UdhZ3MSYm3H:b-1r4h7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L83)):*
> ```Go
> 	"sql.hints.statement_hints_cache_size",
> 	"number of hint entries to store in the LRU",
> 	1024,
> ```

nit: we should add validation here, perhaps non-negative int?
___
*[`pkg/sql/hints/hint_cache.go` line 207 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRcctRCGLb4svkc4Aa:-OaRcctRCGLb4svkc4Ab:b-9wkebj) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L207)):*
> ```Go
> ) (statementHash int64, err error) {
> 	// The rangefeed is set up on the secondary index on the hash column.
> 	dirs := []catenumpb.IndexColumn_Direction{catenumpb.IndexColumn_ASC}
> ```

nit: if this escapes, we could make it a global singleton to avoid an allocation.

Perhaps we could reuse the scratch space for `keyVals` too, but it's not immediately clear to me whether we can have any concurrency of `onValue` calls from the range feed, maybe leave a TODO for that.
___
*[`pkg/sql/hints/hint_cache.go` line 258 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdM3CDDUIitQCFIm5:-OaRdM3CDDUIitQCFIm6:b-w5emco) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L258)):*
> ```Go
> 	ctx context.Context, statementFingerprint string,
> ) (hints []StatementHint) {
> 	hash := fnv.New64()
> ```

nit: on a quick search for differences between fnv1 and fnv1a, the latter seems to be preferred, so maybe use `New64a`?
___
*[`pkg/sql/hints/hint_cache.go` line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdyKX7Yk3XEjPc8Wz:-OaRdyKX7Yk3XEjPc8X-:b-uncjfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L276)):*
> ```Go
> 		}
> 	}
> 	var e interface{}
> ```

nit: do we need this declaration?
___
*[`pkg/sql/hints/hint_cache.go` line 389 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRfKSq49JjSeqwh2vG:-OaRfKSq49JjSeqwh2vH:b-2htpc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L389)):*
> ```Go
> 
> 	// fingerprints and hints have the same length. fingerprints[i] is the
> 	// statement fingerprint to which hints[i] applies.
> ```

nit: IIUC a single fingerprint can have multiple hints (valid use case), and also multiple fingerprints can end up in the same `cacheEntry` due to cache collision, right? It'd be nice to include these details into the comment.

Additionally, given that fingerprint can be of non-trivial length, it might be nice to de-duplicate them. So we could have a structure like
```
  // unique fingerprints
  fingerprints []string
  // hintsIndices[i] contains indices into 'hints' that correspond to the fingerprint 'fingerprints[i]'
  hintsIndices [][]int
  hints []StatementHint
```
This way, if we expect multiple hints per fingerprint to be somewhat frequent, we'll reduce memory footprint as well as speed up the search in `getMatchingHints` a bit. Maybe leave a TODO for this.
___
*[`pkg/sql/hints/statement_hint.go` line 6 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRXiAU5KUe-If0JL-Y:-OaRXiAU5KUe-If0JL-Z:b-s1zp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b0bd14c49b4f8e202bfe486db4f842b2e423b36/pkg/sql/hints/statement_hint.go#L6)):*
> ```Go
> // included in the /LICENSE file.
> 
> package hints
> ```

nit: we should mark `sql/hints` package explicitly as owned by Queries.
___
*[`pkg/sql/catalog/systemschema/system.go` line 5467 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRZGcT3u_q2KS3oI5z:-OaRZGcT3u_q2KS3oI6-:bea3dc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/catalog/systemschema/system.go#L5467)):*
> ```Go
> 			},
> 			descpb.IndexDescriptor{
> 				Name:                "hash_idx",
> ```

nit: not sure if this has been discussed elsewhere, but should this secondary index be covering? Or STORING some extra columns?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/catalog/systemschema/system.go` line 5467 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRZGcT3u_q2KS3oI5z:-OaTGN1X08nhB7IBdxgN:b-1yuphz) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/catalog/systemschema/system.go#L5467)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: not sure if this has been discussed elsewhere, but should this secondary index be covering? Or STORING some extra columns?
</blockquote></details>

The expectation is that this index will be used to:
1. Locate the (few) rows matching a particular hash/fingerprint, in which case the index join is cheap.
2. Drive a rangefeed that listens for the set of all hashes present in the table.

Using a non-storing index balances between quickly accessing hints and duplicating table data. It also reduces the amount of memory needed for the rangefeed. Also, would you be interested in taking a look at the design doc? I've been trying to keep it updated as the actual implementation evolves.
___
*[`pkg/sql/hints/hint_cache.go` line 55 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRbNdoATqph43n836D:-OaTICZIAPQk_J_N9jfV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L55)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it'd be good to mention what are the "key" and the "value" in this map.
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 83 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaR_qFc5UdhZ3MSYm3G:-OaTIQ952nZau61OgM8t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L83)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should add validation here, perhaps non-negative int?
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 207 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRcctRCGLb4svkc4Aa:-OaVqkY_A8_PHYvficIP:bsfez2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L207)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: if this escapes, we could make it a global singleton to avoid an allocation.

Perhaps we could reuse the scratch space for `keyVals` too, but it's not immediately clear to me whether we can have any concurrency of `onValue` calls from the range feed, maybe leave a TODO for that.
</blockquote></details>

Looks like we're smart enough to avoid the allocations in both cases if I'm reading correctly:
https://gist.github.com/DrewKimball/6c599008ef54da9a3dbe66de4e155d06
___
*[`pkg/sql/hints/hint_cache.go` line 258 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdM3CDDUIitQCFIm5:-OaTJbqt9dFCSd06S4hI:bjfoxad) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: on a quick search for differences between fnv1 and fnv1a, the latter seems to be preferred, so maybe use `New64a`?
</blockquote></details>

Hm, this hash has to match what's in the system table, so we'll have to migrate that in order to change it here. Do you know how involved that would be at this point?
___
*[`pkg/sql/hints/hint_cache.go` line 276 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdyKX7Yk3XEjPc8Wz:-OaTK-4s4qb-qvViZYIl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L276)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we need this declaration?
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 389 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRfKSq49JjSeqwh2vG:-OaVpJsZ5OMSdLdMGXqa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L389)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: IIUC a single fingerprint can have multiple hints (valid use case), and also multiple fingerprints can end up in the same `cacheEntry` due to cache collision, right? It'd be nice to include these details into the comment.

Additionally, given that fingerprint can be of non-trivial length, it might be nice to de-duplicate them. So we could have a structure like
```
  // unique fingerprints
  fingerprints []string
  // hintsIndices[i] contains indices into 'hints' that correspond to the fingerprint 'fingerprints[i]'
  hintsIndices [][]int
  hints []StatementHint
```
This way, if we expect multiple hints per fingerprint to be somewhat frequent, we'll reduce memory footprint as well as speed up the search in `getMatchingHints` a bit. Maybe leave a TODO for this.
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 237 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRisPPDv8GWU7PX5VU:-OaTIljVCvn0s65yW8ve:bfw03dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dd47b4ac71506d17e1b811332d5a3009854c477/pkg/sql/hints/hint_cache.go#L237)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Just double-checking that we expect that `onError` and `onUpdate` handlers are called from the same goroutine, right? (I think so based on a quick look at the code.)
</blockquote></details>

That's right - the cluster settings watcher does something similar ([here](https://github.com/cockroachdb/cockroach/blob/8387e9ccb0960b5c207df74f8a8abcc979715eeb/pkg/server/settingswatcher/settings_watcher.go#L190) and [here](https://github.com/cockroachdb/cockroach/blob/8387e9ccb0960b5c207df74f8a8abcc979715eeb/pkg/server/settingswatcher/settings_watcher.go#L274)).
___
*[`pkg/sql/hints/hint_cache.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlqS1-MLC9QbnZErZ:-OaTHOul4EPrV8TArDOh:b-cgldhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L40)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Should we somehow handle the internal queries in a special way? Do we foresee ever adding hints for them? It sounds like with the current approach we expect them to not be found in `hintedHashes` (right?), so maybe the overhead of that check is negligible.
</blockquote></details>

The current plan is to not apply these hints to internal queries at all in V0, and possibly add a toggle for this behavior in some future version. While I can imagine scenarios where we'd want to have this for internal queries, it doesn't seem worth the trouble for now.
___
*[`pkg/sql/hints/hint_cache.go` line 444 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlSj8AgE0ZKje3TK5:-OaVpSvO3cIMsj0df2x2:b-mgtclb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L444)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I have a somewhat general question. Do we want for hints to apply across databases? It seems like right now if we happen to have a hash match and then the stmt fingerprint matches, we'll apply all hints, regardless of the DB, and perhaps that's intended.
</blockquote></details>

The thinking is that we want to start by applying hints broadly, and then we can offer the ability for users to filter hint applications by DB or whatever else later. If we started with per-DB behavior, it might be more difficult to "widen" things in a later version.
___
*[`pkg/sql/hints/statement_hint.go` line 6 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRXiAU5KUe-If0JL-Y:-OaVq5FMBL7ENsSUD2R9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b0bd14c49b4f8e202bfe486db4f842b2e423b36/pkg/sql/hints/statement_hint.go#L6)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should mark `sql/hints` package explicitly as owned by Queries.
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache_test.go` line 36 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRkM7zBH7IauZTyBS0:-OaTFjrK7b2SpiWKYHba:b-vjyogb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache_test.go#L36)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: why are some of these tests explicitly choose to not use randomized test tenant? I'd expect that things should just work, modulo perhaps a more complicated setup where we need to do something (like updating SystemOnly cluster settings) against the system tenant explicitly.
</blockquote></details>

I hadn't devoted much thought to it, TBH. Good catch though, I've added a helper to set these settings directly on the system tenant so that we can use the randomized app tenant.
___
*[`pkg/sql/hints/hint_cache_test.go` line 257 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRmaZVCshVJ6V665Hr:-OaTGIXr4JxUr9d4cP3f:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache_test.go#L257)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: missing period.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> @yuzefovich reviewed 4 of 11 files at r11, 7 of 9 files at r12, 5 of 5 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/catalog/systemschema/system.go` line 5467 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRZGcT3u_q2KS3oI5z:-OaWM_-M7ZwbrRUMD8FM:b-x8u7se) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/catalog/systemschema/system.go#L5467)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

The expectation is that this index will be used to:
1. Locate the (few) rows matching a particular hash/fingerprint, in which case the index join is cheap.
2. Drive a rangefeed that listens for the set of all hashes present in the table.

Using a non-storing index balances between quickly accessing hints and duplicating table data. It also reduces the amount of memory needed for the rangefeed. Also, would you be interested in taking a look at the design doc? I've been trying to keep it updated as the actual implementation evolves.
</blockquote></details>

Yeah, I'd be interested to take a look, thanks!
___
*[`pkg/sql/hints/hint_cache.go` line 207 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRcctRCGLb4svkc4Aa:-OaWN9RU40qdhJaWKMRF:b-drmgx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L207)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Looks like we're smart enough to avoid the allocations in both cases if I'm reading correctly:
https://gist.github.com/DrewKimball/6c599008ef54da9a3dbe66de4e155d06
</blockquote></details>

Nice, thanks for checking.
___
*[`pkg/sql/hints/hint_cache.go` line 258 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdM3CDDUIitQCFIm5:-OaWNJUv0KkbN73bUM1N:bn0ec4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L258)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hm, this hash has to match what's in the system table, so we'll have to migrate that in order to change it here. Do you know how involved that would be at this point?
</blockquote></details>

Which system table are you referring to?

If `system.statement_statistics`, then we do need to match the calculation, and the migration doesn't seem worth it. It's a bit non-trivial to see that our hash computation here must match what we do in `appstatspb.ConstructStatementFingerprintID`. We should add a comment here about that. Also it looks like we include `implicitTxn` and `database` into the fingerprint hash in `appstatspb`, is it a problem?
- Should we also use `util.FNV64` package instead of `fnv` just to be safe (since that's what we use in `ConstructStatementFingerprintID`)?

If `system.statement_hints`, i.e. our hash computation is independent of whatever we do in other system tables, then we could change the computation with no migration, since we haven't inserted any hints yet, right?

(Perhaps I need to just read the design doc :sweat_smile: )
___
*[`pkg/sql/hints/hint_cache.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlqS1-MLC9QbnZErZ:-OaWMmbR21hLgNPzvOc8:b-36miz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L40)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

The current plan is to not apply these hints to internal queries at all in V0, and possibly add a toggle for this behavior in some future version. While I can imagine scenarios where we'd want to have this for internal queries, it doesn't seem worth the trouble for now.
</blockquote></details>

Ack, makes sense. The hints not being applied to internal queries will be implemented at a different level, right?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @yuzefovich reviewed 5 of 5 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-OaaWtzG-taeuhs5LDvH:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/hints/hint_cache.go` line 258 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdM3CDDUIitQCFIm5:-OaaSW5LCndWVH1lcnxE:b6j3yig) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Which system table are you referring to?

If `system.statement_statistics`, then we do need to match the calculation, and the migration doesn't seem worth it. It's a bit non-trivial to see that our hash computation here must match what we do in `appstatspb.ConstructStatementFingerprintID`. We should add a comment here about that. Also it looks like we include `implicitTxn` and `database` into the fingerprint hash in `appstatspb`, is it a problem?
- Should we also use `util.FNV64` package instead of `fnv` just to be safe (since that's what we use in `ConstructStatementFingerprintID`)?

If `system.statement_hints`, i.e. our hash computation is independent of whatever we do in other system tables, then we could change the computation with no migration, since we haven't inserted any hints yet, right?

(Perhaps I need to just read the design doc :sweat_smile: )
</blockquote></details>

Thanks for starting a slack thread for this question. Given that we've already shipped a beta, we'd need to add a migration to change the hash function, so we can ignore this comment.
___
*[`pkg/sql/hints/hint_cache.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlqS1-MLC9QbnZErZ:-OaaWwwHEiNwYhTbFDGl:b5cx54j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L40)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack, makes sense. The hints not being applied to internal queries will be implemented at a different level, right?
</blockquote></details>

Still curious about this question but it's non-blocking.
___
*[`pkg/sql/hints/hint_cache_test.go` line 120 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaW0Pl0rwlMi0cXU9v:-OaaW0Pl0rwlMi0cXU9w:b-f9ha2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache_test.go#L120)):*
> ```Go
> 		return nil
> 	})
> 	initialReads := hc.TestingNumTableReads()
> ```

nit: we need this because of internal queries? Consider adding a comment quick for why we need this offset.
___
*[`pkg/sql/hints/hint_cache.go` line 54 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaSk7h3y_dAZfOtn2P:-OaaSk7h3y_dAZfOtn2Q:bjk2pal) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache.go#L54)):*
> ```Go
> 		// quickly determine whether there are any hints for a given statement,
> 		// minimizing the overhead for non-hinted statements.
> 		hintedHashes map[int64]int
> ```

nit: add a TODO to add memory monitoring (as mentioned in the design doc).
___
*[`pkg/sql/hints/hint_cache.go` line 324 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaTmD55Dcpg96VioJm:-OaaTmD55Dcpg96VioJn:b-oh25m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache.go#L324)):*
> ```Go
> 		// The plan hints were evicted from the cache. Retrieve them from the
> 		// database and add them to the cache.
> 		return c.addCacheEntryLocked(ctx, statementHash, statementFingerprint)
> ```

nit: it might be nice to add `mu.AssertHeld()` at the top of both `*Locked` methods.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
TFYR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-OaaoAxaEmPDDajhqsL4:b-814cj3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/hints/hint_cache.go` line 258 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRdM3CDDUIitQCFIm5:-OaalGTI96UMZTxoYm5U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/160d133be5b9a0c88bc8eae2d531b80a1b9a7c2e/pkg/sql/hints/hint_cache.go#L258)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks for starting a slack thread for this question. Given that we've already shipped a beta, we'd need to add a migration to change the hash function, so we can ignore this comment.
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaRlqS1-MLC9QbnZErZ:-OaalHnM18e2BU6ehCQ6:bvrim4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a876e65240f8804ffdd6da00e3cc06b0117ac896/pkg/sql/hints/hint_cache.go#L40)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Still curious about this question but it's non-blocking.
</blockquote></details>

Yes, we'll decide whether or not to look up + apply hints in the connExecutor.
___
*[`pkg/sql/hints/hint_cache.go` line 54 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaSk7h3y_dAZfOtn2P:-Oaan9d93RbKtVfE-Gte:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache.go#L54)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: add a TODO to add memory monitoring (as mentioned in the design doc).
</blockquote></details>

Done.
___
*[`pkg/sql/hints/hint_cache.go` line 324 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaTmD55Dcpg96VioJm:-OaanF2F4dM3MAoDHhWG:b-ydonir) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache.go#L324)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it might be nice to add `mu.AssertHeld()` at the top of both `*Locked` methods.
</blockquote></details>

Done. Good idea.
___
*[`pkg/sql/hints/hint_cache_test.go` line 120 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OaaW0Pl0rwlMi0cXU9v:-Oaal_9R8Bgoxwpad8QC:bu3votc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aeccfaed98b81527e79fdfba8854daf16dc93a6/pkg/sql/hints/hint_cache_test.go#L120)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we need this because of internal queries? Consider adding a comment quick for why we need this offset.
</blockquote></details>

Hm, that's a good point, we don't actually need the offset. Done.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Awesome job. Some questions but no blockers.

(Don't forget to update the table name in the PR summary.)

> @michae2 reviewed 1 of 12 files at r7, 12 of 12 files at r17, 9 of 9 files at r18, 1 of 1 files at r19, 5 of 5 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-OafS0_32NkdOVSR3IjX:bs602t3)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/hints/hint_cache.go` line 265 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oa2HghRBeQJebj3Zs_f:-OafM1q79awfinyWa2nt:bu7i25p) ([raw file](https://github.com/cockroachdb/cockroach/blob/55c7de1356bdcc394d33ef777c1c29dc8d0aa00e/pkg/sql/hints/hint_cache.go#L265)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'd argue the current state is more defensive, since it means we'll fall back to the LRU cache and read any hints from the table. What do you think?
</blockquote></details>

Make sense.
___
*[`-- commits` line 14 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oaf8iTv6ATXdkuZ2Q_o:-Oaf8iTv6ATXdkuZ2Q_p:bddyrm3):*
typo: table name is wrong
___
*[`-- commits` line 39 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafONv83xsaVXy4jS7S:-OafONv83xsaVXy4jS7T:bcqyb8e):*
I'm thinking about the case where system.statement_hints is unavailable, but we still want most user queries to go through. I wonder if we need an additional fallback path where we run without any hints (or with only the hints currently in cache).

(Also, this makes me wonder what happens when system.table_statistics is unavailable... :grimacing:)
___
*[`pkg/sql/hints/hint_cache.go` line 55 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafBXk7EBsxyTdJ5iqd:-OafBXk7EBsxyTdJ5iqe:b4yw4nx) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L55)):*
> ```Go
> 		// quickly determine whether there are any hints for a given statement,
> 		// minimizing the overhead for non-hinted statements.
> 		hintedHashes map[int64]int
> ```

Should this be uint64? I see that [Hash64.Sum64](https://pkg.go.dev/hash#Hash64) returns a uint64.
___
*[`pkg/sql/hints/hint_cache.go` line 79 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafC2mQ8bh_Gcu0g6Mv:-OafC2mQ8bh_Gcu0g6Mw:bttjitx) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L79)):*
> ```Go
> // In general this should be larger than the number of active statement
> // fingerprints with hints.
> var cacheSize = settings.RegisterIntSetting(
> ```

Would it be worth making this metamorphic in tests? Maybe on master?
___
*[`pkg/sql/hints/hint_cache.go` line 381 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafMq-A7IaZIIr0RC-G:-OafMda-AMT64itSFmcL:b-pgcvz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L381)):*
> ```Go
> 		entry.fingerprints = append(entry.fingerprints, fingerprint)
> 		entry.hints = append(entry.hints, hint)
> 	}
> ```

Should we be updating the count in hintedHashes after this query? I guess the rangefeed should keep it accurate...
___
*[`pkg/sql/hints/hint_cache.go` line 408 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafH7ygE-dyUALGEUpR:-OafH7ygE-dyUALGEUpS:bx6ziva) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L408)):*
> ```Go
> 	for i := range entry.hints {
> 		if entry.fingerprints[i] == statementFingerprint {
> 			// TODO(drewk): we could do something smarter here to avoid allocations.
> ```

commentary: we could `ORDER BY fingerprint` in the query and then use a slice (might also want to change the secondary index to include fingerprint)

(no need to do this now)
___
*[`pkg/sql/hints/hint_cache.go` line 134 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafQJqF0m6FL1TImi5O:-OafQJqF0m6FL1TImi5P:bm8a7vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ff891eab0782ab006db7b8d344a6dfd5a9d66cb/pkg/sql/hints/hint_cache.go#L134)):*
> ```Go
> 		return errors.Wrap(stop.ErrUnavailable, "failed to retrieve initial statement hints")
> 
> 	case <-ctx.Done():
> ```

I wonder if we should wrap ctx with a context.WithTimeout just to be safe?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
A list of additional changes (that we've discussed in meetings / on slack):
1. I discovered that we need to pass a non-zero buffer value to `rangefeedcache.NewWatcher` (see [here](https://cockroachlabs.slack.com/archives/C07T05244KZ/p1759530812328979)) to prevent spurious rangefeed restarts. I added a metamorphic constant for that.
2. After fixing the above, test failures revealed that rangefeed events can be duplicated, so the counting `hintedHashes` map doesn't work. I've changed the cache so that it reads from the table for deletion events, with a `pendingRefreshes` map to ensure there's only one refresh active for any given hash.

I'll spend some more time making the changes for (2) more readable, but the tests are now passing reliably.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-Oauvxid1Rz6_rgRXH6W:bnt2j4v)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner, @michae2, and @yuzefovich)

___
*[`-- commits` line 14 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-Oaf8iTv6ATXdkuZ2Q_o:-Ob1aqUb7fK9cohxiZuC:b-896fix):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

typo: table name is wrong
</blockquote></details>

Done.
___
*[`-- commits` line 39 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafONv83xsaVXy4jS7S:-OaumfFn1Ga1VAW77Wyp:b-qa55c1):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I'm thinking about the case where system.statement_hints is unavailable, but we still want most user queries to go through. I wonder if we need an additional fallback path where we run without any hints (or with only the hints currently in cache).

(Also, this makes me wonder what happens when system.table_statistics is unavailable... :grimacing:)
</blockquote></details>

If there's an error fetching hints from the DB, we log the error and remove the corresponding entry from the cache. The current query (and any others that were waiting on that cache entry) will proceed with no hints applied. The table stats cache does something similar, and in fact is what I based that logic on. I'll add this explicitly to the design doc.
___
*[`pkg/sql/hints/hint_cache.go` line 55 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafBXk7EBsxyTdJ5iqd:-OaujCjrEeVmCAMk_1SI:b-kbspwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L55)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be uint64? I see that [Hash64.Sum64](https://pkg.go.dev/hash#Hash64) returns a uint64.
</blockquote></details>

We're matching the `fnv64` builtin function in the table, which casts to `int64`.
___
*[`pkg/sql/hints/hint_cache.go` line 79 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafC2mQ8bh_Gcu0g6Mv:-OaujO8VB-WM247NQ2AN:b-egv461) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L79)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Would it be worth making this metamorphic in tests? Maybe on master?
</blockquote></details>

That's a good idea. Done.
___
*[`pkg/sql/hints/hint_cache.go` line 381 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafMq-A7IaZIIr0RC-G:-OaukmOlCAHU8Xtlzd7D:bk2sri7) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L381)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should we be updating the count in hintedHashes after this query? I guess the rangefeed should keep it accurate...
</blockquote></details>

The rangefeed that's maintaining `hintedHashes` will generally lag the current timestamp by a few seconds, so we expect the counts to differ for a short time (e.g. when going from one hint to two hints). If we did "correct" the count at this point using what we read from the DB at the current timestamp, the lagging updates applied by the rangefeed would cause the count to become too large or too small.

Note that we can always restart the rangefeed to re-initialize the map (we should add a builtin for this).
___
*[`pkg/sql/hints/hint_cache.go` line 408 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafH7ygE-dyUALGEUpR:-Oaulqxo4ZJojVaKl9Mm:buhk8yd) ([raw file](https://github.com/cockroachdb/cockroach/blob/21e0aca3233dd210ccfb67f5063b1589104d06f5/pkg/sql/hints/hint_cache.go#L408)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

commentary: we could `ORDER BY fingerprint` in the query and then use a slice (might also want to change the secondary index to include fingerprint)

(no need to do this now)
</blockquote></details>

Nice, I'll add that to the comment.
___
*[`pkg/sql/hints/hint_cache.go` line 134 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-OafQJqF0m6FL1TImi5O:-Oaur8qBAoJfkqKHGmZa:b-x1rm40) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ff891eab0782ab006db7b8d344a6dfd5a9d66cb/pkg/sql/hints/hint_cache.go#L134)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I wonder if we should wrap ctx with a context.WithTimeout just to be safe?
</blockquote></details>

Good idea. I added a one minute timeout and a log message every 5 seconds.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Truly a pleasure to read this code!

> @michae2 reviewed 12 of 12 files at r21, 8 of 10 files at r22, 13 of 13 files at r25, 10 of 10 files at r26, 2 of 2 files at r27, 5 of 5 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-ObENkG_EKI3SypnBTyw:b-3125q8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @yuzefovich)

___
*[`pkg/sql/hints/hint_cache.go` line 151 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEKuzF9-00Xi8WhBPZ:-ObEKuzF9-00Xi8WhBP_:b-72h4df) ([raw file](https://github.com/cockroachdb/cockroach/blob/531ea6866be593601c0874c9907924b6952050ca/pkg/sql/hints/hint_cache.go#L151)):*
> ```Go
> 	// so it is OK to return early if the scan takes too long or has to retry.
> 	// If we do return early, statements will simply skip hintedHashes and check
> 	// hintCache until the initial scan finishes.
> ```

In the return-early scenario, for queries without any hints, will we cache an empty slice after querying statement_hints for the first time?

If so, that seems perfectly fine (except for the memory usage). Maybe it's not worth adding this blocking here, and just letting the query always be async?
___
*[`pkg/sql/hints/hint_cache.go` line 158 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEM9r-DNadrxv9Nrr0:-ObEM0l3BHhv60qBZO45:b-4u6xab) ([raw file](https://github.com/cockroachdb/cockroach/blob/531ea6866be593601c0874c9907924b6952050ca/pkg/sql/hints/hint_cache.go#L158)):*
> ```Go
> 			logTimer.Reset(time.Second * 5)
> 
> 		case <-time.After(time.Minute):
> ```

1 minute seems really long IMO, and not worth waiting for. (I'm imagining this multiplied by all tests and all CI runs and all customers...)
___
*[`pkg/sql/hints/hint_cache.go` line 76 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObENNcg2yx1vDDJKi-A:-ObENNcg2yx1vDDJKi-B:b-o22daa) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa283b6e051ee9bcaae9819a9bd44939a2c6a6e0/pkg/sql/hints/hint_cache.go#L76)):*
> ```Go
> 		// from the system table. Note that this count only includes reads used for
> 		// hintCache, not for hintedHashes.
> 		numInternalQueries int
> ```

I could imagine this becoming a metric one day.
___
*[`pkg/sql/hints/hint_cache.go` line 260 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEAUzrBHx1i5uFkWXP:-ObEAUzrBHx1i5uFkWXQ:b-8e6oig) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eadc51e467b09dd5cddc9c371e8a441ef3d024e/pkg/sql/hints/hint_cache.go#L260)):*
> ```Go
> 		// TODO(drewk): an insertion event implies on its own that there is at least
> 		// one hint at that timestamp, so we could skip the DB read for insertions.
> 		refreshTS := c.clock.Now()
> ```

Why is this c.clock.Now() rather than ev.ts? Just so we only query a hash once per incremental update? Could it cause us to miss another incremental update that happens concurrently with a timestamp > ev.ts but < c.clock.Now()?

I wonder if it would be better to iterate the update.Events in reverse order? (Assuming they're ordered by time.)
___
*[`pkg/sql/hints/hint_cache.go` line 462 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEHUO235K1gXl1irxU:-ObEHUO235K1gXl1irxV:b-wjs9h2) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eadc51e467b09dd5cddc9c371e8a441ef3d024e/pkg/sql/hints/hint_cache.go#L462)):*
> ```Go
> 	if err != nil {
> 		// Remove the cache entry so that the next caller retries the query.
> 		c.mu.hintCache.Del(statementHash)
> ```

Will concurrent waiters see zero hints? (That seems fine, just checking if I understand.)


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
Thanks for taking another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-:-ObEYuKP47QEV9qp5Dfk:b-pvxpm6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @michae2, and @yuzefovich)

___
*[`pkg/sql/hints/hint_cache.go` line 260 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEAUzrBHx1i5uFkWXP:-ObESV88AVaKItyHPZa-:b-7bvzkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eadc51e467b09dd5cddc9c371e8a441ef3d024e/pkg/sql/hints/hint_cache.go#L260)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why is this c.clock.Now() rather than ev.ts? Just so we only query a hash once per incremental update? Could it cause us to miss another incremental update that happens concurrently with a timestamp > ev.ts but < c.clock.Now()?

I wonder if it would be better to iterate the update.Events in reverse order? (Assuming they're ordered by time.)
</blockquote></details>

If the refreshing read is happening at a later timestamp than the other incremental updates, it will observe the effects of those updates, so no need to do any more work for them. I added a bit to the comment.

Note that we could use `ev.ts` with AOST or something, but doing things the current way reduces the amount of work if there are multiple events clustered together, and also avoids any (admittedly unlikely) issues with reads before the GC threshold.
___
*[`pkg/sql/hints/hint_cache.go` line 462 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEHUO235K1gXl1irxU:-ObEVtZW9pkveB8KqBk4:bamgtwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eadc51e467b09dd5cddc9c371e8a441ef3d024e/pkg/sql/hints/hint_cache.go#L462)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Will concurrent waiters see zero hints? (That seems fine, just checking if I understand.)
</blockquote></details>

If there's an error, concurrent waiters on `entry.waitCond` will see zero hints, yes. Any later goroutines will retry the DB read, since the entry was removed from the cache.
___
*[`pkg/sql/hints/hint_cache.go` line 151 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEKuzF9-00Xi8WhBPZ:-ObEP8GP6a9_uNbmLVcs:b-gmh7xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/531ea6866be593601c0874c9907924b6952050ca/pkg/sql/hints/hint_cache.go#L151)):*
>In the return-early scenario, for queries without any hints, will we cache an empty slice after querying statement\_hints for the first time?

Yes, we'll just fall back to always using `hintCache` instead of skipping it for non-hinted statements until `hintedHashes` is initialized.

>If so, that seems perfectly fine (except for the memory usage). Maybe it's not worth adding this blocking here, and just letting the query always be async?

Yeah, removing the blocking SGTM. It'll cut down on complexity and reduce the chance of test slowdowns.
___
*[`pkg/sql/hints/hint_cache.go` line 158 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObEM9r-DNadrxv9Nrr0:-ObEPVND4evOfV8423g2:b8m8ivl) ([raw file](https://github.com/cockroachdb/cockroach/blob/531ea6866be593601c0874c9907924b6952050ca/pkg/sql/hints/hint_cache.go#L158)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

1 minute seems really long IMO, and not worth waiting for. (I'm imagining this multiplied by all tests and all CI runs and all customers...)
</blockquote></details>

Yeah, I wouldn't expect to hit the timeout even for clusters with many hints, but I agree that it's simplest to just remove the blocking.
___
*[`pkg/sql/hints/hint_cache.go` line 76 at r28](https://reviewable.io/reviews/cockroachdb/cockroach/153995#-ObENNcg2yx1vDDJKi-A:-ObEWeg4C5fwmSQEtBi1:b-1dskdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa283b6e051ee9bcaae9819a9bd44939a2c6a6e0/pkg/sql/hints/hint_cache.go#L76)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I could imagine this becoming a metric one day.
</blockquote></details>

That could be useful. I added a TODO.


<!-- Sent from Reviewable.io -->
